File: src/smart_signal/parser.py
"""JSON parsing helpers for Binance Smart Signal payloads."""

from __future__ import annotations

import math
import re
from dataclasses import dataclass
from typing import Any


POSITION_KEYS = {
    "positionusd",
    "positionusdt",
    "position_usdt",
    "position",
    "positionvalue",
}
AVG_ENTRY_KEYS = {
    "avgentryprice",
    "avg_entry_price",
    "entryprice",
    "averageentryprice",
}
LONG_ALIASES = {"long", "buy", "bull", "bullish"}
SHORT_ALIASES = {"short", "sell", "bear", "bearish"}
NUMERIC_TOKEN_RE = re.compile(r"^\s*([+-]?\d+(?:\.\d+)?)\s*([KMBT]?)\s*$", re.IGNORECASE)
FLAT_KEY_RE = re.compile(r"^(long|short)[_\-]?(.*)$", re.IGNORECASE)

MULTIPLIERS = {
    "": 1.0,
    "K": 1_000.0,
    "M": 1_000_000.0,
    "B": 1_000_000_000.0,
    "T": 1_000_000_000_000.0,
}


@dataclass(slots=True)
class SideSnapshot:
    side: str
    position_qty: float | None = None
    position_usdt: float | None = None
    avg_entry_price: float | None = None
    source_path: str | None = None


@dataclass(slots=True)
class ParsedSignal:
    long: SideSnapshot
    short: SideSnapshot

    def as_rows(self, *, ts_utc: str, symbol: str, cohort: str, raw_json: str | None) -> list[dict]:
        return [
            {
                "ts_utc": ts_utc,
                "symbol": symbol,
                "cohort": cohort,
                "side": side_snapshot.side,
                "position_qty": side_snapshot.position_qty,
                "position_usdt": side_snapshot.position_usdt,
                "avg_entry_price": side_snapshot.avg_entry_price,
                "raw_json": raw_json,
            }
            for side_snapshot in (self.long, self.short)
        ]


def build_position_usdt(quantity: Any, current_price: Any) -> float | None:
    normalized_qty = normalize_number(quantity)
    normalized_price = normalize_number(current_price)
    if normalized_qty is None or normalized_price is None:
        return None
    return normalized_qty * normalized_price


def normalize_number(value: Any) -> float | None:
    """Normalize Binance UI numeric strings such as 951.89M or 2.41B."""
    if value is None:
        return None
    if isinstance(value, bool):
        return None
    if isinstance(value, (int, float)):
        if isinstance(value, float) and (math.isnan(value) or math.isinf(value)):
            return None
        return float(value)

    text = str(value).strip()
    if not text or text in {"--", "N/A", "null", "None"}:
        return None

    compact = text.replace(",", "").replace("$", "").replace("USDT", "").replace("USD", "").strip()
    match = NUMERIC_TOKEN_RE.match(compact)
    if not match:
        return None

    number, suffix = match.groups()
    return float(number) * MULTIPLIERS[suffix.upper()]


def parse_signal_payload(payload: Any) -> ParsedSignal:
    """Extract long/short metrics from loosely structured Smart Signal payloads."""
    overview_parsed = parse_overview_payload(payload)
    if overview_parsed is not None:
        return overview_parsed

    snapshots = {
        "long": SideSnapshot(side="long"),
        "short": SideSnapshot(side="short"),
    }
    _walk_payload(payload, snapshots, "$")
    return ParsedSignal(long=snapshots["long"], short=snapshots["short"])


def parse_overview_payload(payload: Any, *, current_price: Any | None = None, cohort: str = "trader") -> ParsedSignal | None:
    """Parse Binance Smart Signal overview payload for trader or whale cohorts."""
    if not isinstance(payload, dict):
        return None

    data = payload.get("data", payload)
    if not isinstance(data, dict):
        return None

    prefix = "Traders" if cohort == "trader" else "Whales"
    long_qty_key = f"long{prefix}Qty"
    short_qty_key = f"short{prefix}Qty"
    long_avg_key = f"long{prefix}AvgEntryPrice"
    short_avg_key = f"short{prefix}AvgEntryPrice"

    if long_qty_key not in data and short_qty_key not in data:
        return None

    return ParsedSignal(
        long=SideSnapshot(
            side="long",
            position_qty=normalize_number(data.get(long_qty_key)),
            position_usdt=build_position_usdt(data.get(long_qty_key), current_price),
            avg_entry_price=normalize_number(data.get(long_avg_key)),
            source_path=f"$.data.{long_qty_key}",
        ),
        short=SideSnapshot(
            side="short",
            position_qty=normalize_number(data.get(short_qty_key)),
            position_usdt=build_position_usdt(data.get(short_qty_key), current_price),
            avg_entry_price=normalize_number(data.get(short_avg_key)),
            source_path=f"$.data.{short_qty_key}",
        ),
    )


def _walk_payload(node: Any, snapshots: dict[str, SideSnapshot], path: str) -> None:
    if isinstance(node, dict):
        _extract_nested_side_objects(node, snapshots, path)
        _extract_flattened_side_keys(node, snapshots, path)
        _extract_side_record(node, snapshots, path)

        for key, value in node.items():
            _walk_payload(value, snapshots, f"{path}.{key}")
    elif isinstance(node, list):
        for index, item in enumerate(node):
            _walk_payload(item, snapshots, f"{path}[{index}]")


def _extract_nested_side_objects(node: dict[str, Any], snapshots: dict[str, SideSnapshot], path: str) -> None:
    for side in ("long", "short"):
        raw_value = node.get(side)
        if isinstance(raw_value, dict):
            _merge_side_values(snapshots[side], raw_value, f"{path}.{side}")


def _extract_flattened_side_keys(node: dict[str, Any], snapshots: dict[str, SideSnapshot], path: str) -> None:
    grouped: dict[str, dict[str, Any]] = {"long": {}, "short": {}}
    for key, value in node.items():
        if not isinstance(key, str):
            continue
        match = FLAT_KEY_RE.match(key)
        if not match:
            continue
        side, tail = match.groups()
        grouped[side.lower()][tail.lower()] = value

    for side, values in grouped.items():
        if values:
            _merge_side_values(snapshots[side], values, f"{path}.{side}_flat")


def _extract_side_record(node: dict[str, Any], snapshots: dict[str, SideSnapshot], path: str) -> None:
    side = _coerce_side(node.get("side") or node.get("positionSide") or node.get("direction"))
    if side is None:
        return
    _merge_side_values(snapshots[side], node, path)


def _coerce_side(value: Any) -> str | None:
    if value is None:
        return None
    lowered = str(value).strip().lower()
    if lowered in LONG_ALIASES:
        return "long"
    if lowered in SHORT_ALIASES:
        return "short"
    return None


def _merge_side_values(target: SideSnapshot, source: dict[str, Any], path: str) -> None:
    position = _pick_numeric_value(source, POSITION_KEYS)
    avg_entry = _pick_numeric_value(source, AVG_ENTRY_KEYS)

    if position is not None:
        target.position_qty = position
        target.position_usdt = position
        target.source_path = path
    if avg_entry is not None:
        target.avg_entry_price = avg_entry
        target.source_path = path


def _pick_numeric_value(source: dict[str, Any], aliases: set[str]) -> float | None:
    normalized_map = {str(key).replace("-", "").replace("_", "").lower(): value for key, value in source.items()}
    for alias in aliases:
        normalized_alias = alias.replace("-", "").replace("_", "").lower()
        if normalized_alias in normalized_map:
            return normalize_number(normalized_map[normalized_alias])
    return None

File: cli/internal/parser/parser.go
// Package parser provides number normalization and signal parsing for Binance
// Smart Signal payloads. It mirrors the Python smart_signal.parser module.
package parser

import (
	"math"
	"regexp"
	"strings"
)

var numericTokenRE = regexp.MustCompile(`(?i)^\s*([+-]?\d+(?:\.\d+)?)\s*([KMBT]?)\s*$`)

var multipliers = map[string]float64{
	"":  1,
	"K": 1_000,
	"M": 1_000_000,
	"B": 1_000_000_000,
	"T": 1_000_000_000_000,
}

// NormalizeNumber parses Binance UI numeric strings such as "951.89M" or "2.41B".
// Returns nil when the value cannot be parsed.
func NormalizeNumber(value interface{}) *float64 {
	switch v := value.(type) {
	case nil:
		return nil
	case bool:
		return nil
	case float64:
		if math.IsNaN(v) || math.IsInf(v, 0) {
			return nil
		}
		return &v
	case float32:
		f := float64(v)
		return &f
	case int:
		f := float64(v)
		return &f
	case int64:
		f := float64(v)
		return &f
	case string:
		return parseNumericString(v)
	default:
		return nil
	}
}

func parseNumericString(text string) *float64 {
	text = strings.TrimSpace(text)
	if text == "" || text == "--" || text == "N/A" || text == "null" || text == "None" {
		return nil
	}

	compact := strings.ReplaceAll(text, ",", "")
	compact = strings.ReplaceAll(compact, "$", "")
	compact = strings.ReplaceAll(compact, "USDT", "")
	compact = strings.ReplaceAll(compact, "USD", "")
	compact = strings.TrimSpace(compact)

	matches := numericTokenRE.FindStringSubmatch(compact)
	if matches == nil {
		return nil
	}

	number := 0.0
	_, err := parseFloat(matches[1], &number)
	if err {
		return nil
	}

	suffix := strings.ToUpper(matches[2])
	mult, ok := multipliers[suffix]
	if !ok {
		return nil
	}

	result := number * mult
	return &result
}

func parseFloat(s string, out *float64) (int, bool) {
	// simple float parser
	var f float64
	n, err := scanFloat(s, &f)
	if err {
		return 0, true
	}
	*out = f
	return n, false
}

func scanFloat(s string, out *float64) (int, bool) {
	if s == "" {
		return 0, true
	}
	negative := false
	i := 0
	if s[0] == '+' {
		i++
	} else if s[0] == '-' {
		negative = true
		i++
	}

	var intPart float64
	hasDigits := false
	for i < len(s) && s[i] >= '0' && s[i] <= '9' {
		intPart = intPart*10 + float64(s[i]-'0')
		hasDigits = true
		i++
	}

	var fracPart float64
	if i < len(s) && s[i] == '.' {
		i++
		divisor := 10.0
		for i < len(s) && s[i] >= '0' && s[i] <= '9' {
			fracPart += float64(s[i]-'0') / divisor
			divisor *= 10
			hasDigits = true
			i++
		}
	}

	if !hasDigits {
		return 0, true
	}

	result := intPart + fracPart
	if negative {
		result = -result
	}
	*out = result
	return i, false
}

// SideSnapshot holds parsed data for one side (long or short).
type SideSnapshot struct {
	Side         string   `json:"side"`
	PositionQty  *float64 `json:"position_qty"`
	PositionUSDT *float64 `json:"position_usdt"`
	AvgEntryPrice *float64 `json:"avg_entry_price"`
}

// ParsedSignal holds parsed long and short data.
type ParsedSignal struct {
	Long  SideSnapshot `json:"long"`
	Short SideSnapshot `json:"short"`
}

// ParseOverviewPayload parses a Binance Smart Signal overview API response
// for a given cohort ("trader" or "whale").
func ParseOverviewPayload(payload map[string]interface{}, cohort string) *ParsedSignal {
	data, ok := payload["data"]
	if !ok {
		data = payload
	}
	dataMap, ok := data.(map[string]interface{})
	if !ok {
		return nil
	}

	prefix := "Traders"
	if cohort != "trader" {
		prefix = "Whales"
	}

	longQtyKey := "long" + prefix + "Qty"
	shortQtyKey := "short" + prefix + "Qty"
	longAvgKey := "long" + prefix + "AvgEntryPrice"
	shortAvgKey := "short" + prefix + "AvgEntryPrice"

	_, hasLong := dataMap[longQtyKey]
	_, hasShort := dataMap[shortQtyKey]
	if !hasLong && !hasShort {
		return nil
	}

	return &ParsedSignal{
		Long: SideSnapshot{
			Side:         "long",
			PositionQty:  NormalizeNumber(dataMap[longQtyKey]),
			AvgEntryPrice: NormalizeNumber(dataMap[longAvgKey]),
		},
		Short: SideSnapshot{
			Side:         "short",
			PositionQty:  NormalizeNumber(dataMap[shortQtyKey]),
			AvgEntryPrice: NormalizeNumber(dataMap[shortAvgKey]),
		},
	}
}
